File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        //Kadane's Algorithm
        int n=nums.size();
        int sum=0;
        int maxi=INT_MIN;
        for(int i=0;i<n;i++){
            sum+=nums[i];
            if(sum>maxi){
                maxi=sum;
            }

            if(sum<0){ //restart
                sum=0;
            }
        }
        return maxi;
    }
};

File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        //Kadane's Algorithm - Maximum Subarray Sum
        //Intuition: If sum<0, then sum=0
        int sum=0;
        int maxSum=Integer.MIN_VALUE;
        for(int i: nums){
            sum+=i;
            maxSum = Math.max(maxSum,sum);
            if(sum<0){
                sum=0;
            }
        }
        return maxSum;
    }
}